packages/@aws-cdk/aws-msk/lib/cluster.ts

```diff
@@ -0,0 +1,771 @@
+          actions: [
+            'kms:Encrypt',
+            'kms:Decrypt',
+            'kms:ReEncrypt*',
+            'kms:GenerateDataKey*',
+            'kms:CreateGrant',
+            'kms:DescribeKey',
+          ],
+          resources: ['*'],
+          conditions: {
+            StringEquals: {
+              'kms:ViaService': `secretsmanager.${core.Stack.of(this).region}.amazonaws.com`,
+              'kms:CallerAccount': core.Stack.of(this).account,
+            },
+          },
+        }),
+      );
+    }
+    const clientAuthentication = props.clientAuthentication
+      ? {
+        sasl: props.clientAuthentication?.saslProps?.scram
+          ? {
+            scram: {
+              enabled: props.clientAuthentication?.saslProps.scram,
+            },
+          }
+          : undefined,
+        tls: props.clientAuthentication?.tlsProps?.certificateAuthorityArns
+          ? {
+            certificateAuthorityArnList:
+                  props.clientAuthentication?.tlsProps
+                    ?.certificateAuthorityArns,
+          }
+          : undefined,
+      }
+      : undefined;
+
+    const resource = new CfnCluster(this, 'Resource', {
+      clusterName: props.clusterName,
+      kafkaVersion:
+        props.kafkaVersion?.version ?? KafkaVersion.V2_2_1.version,
+      numberOfBrokerNodes:
+        props.numberOfBrokerNodes !== undefined ? props.numberOfBrokerNodes : 1,
+      brokerNodeGroupInfo: {
+        instanceType,
+        clientSubnets: subnetSelection.subnetIds,
+        securityGroups: this.connections.securityGroups.map(
+          (group) => group.securityGroupId,
+        ),
+        storageInfo: {
+          ebsStorageInfo: {
+            volumeSize: volumeSize,
+          },
+        },
+      },
+      encryptionInfo: {
+        encryptionAtRest,
+        encryptionInTransit,
+      },
+      configurationInfo: props.configurationInfo,
+      enhancedMonitoring: props.monitoring?.clusterMonitoringLevel,
+      openMonitoring: openMonitoring,
+      loggingInfo: loggingInfo,
+      clientAuthentication: clientAuthentication,
+    });
+
+    this.clusterName = this.getResourceNameAttribute(
+      core.Fn.select(1, core.Fn.split('/', resource.ref)),
+    );
+    this.clusterArn = resource.ref;
+
+    resource.applyRemovalPolicy(props.removalPolicy, {
+      default: core.RemovalPolicy.RETAIN,
+    });
+  }
+
+  private mskInstanceType(instanceType: ec2.InstanceType): string {
+    return `kafka.${instanceType.toString()}`;
+  }
+
+  /**
+   * Get the ZooKeeper Connection string
+   *
+   * Uses a Custom Resource to make an API call to `describeCluster` using the Javascript SDK
+   *
+   * @param responseField Field to return from API call. eg. ZookeeperConnectString, ZookeeperConnectStringTls
+   * @returns - The connection string to use to connect to the Apache ZooKeeper cluster.
+   */
+  private _zookeeperConnectionString(responseField: string): string {
+    return core.Lazy.stringValue({
```

Comment:
Thats actually not what I meant. This would create a token each time it is called - and at synth time it would still create two constructs with the same id. 

Make sure you add a test for this.

What I meant is this:

```ts
  private _zookeeperConnectionString(responseField: string): string {
    if (!this._clusterDescription) {
      this._clusterDescription = new cr.AwsCustomResource(this, 'ZookeeperConnect', {
        onUpdate: {
          service: 'Kafka',
          action: 'describeCluster',
          parameters: {
            ClusterArn: this.clusterArn,
          },
          physicalResourceId: cr.PhysicalResourceId.of(
            'ZooKeeperConnectionString',
          ),
        },
        policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
          resources: [this.clusterArn],
        }),
      });
    }
    return this._clusterDescription.getResponseField(`ClusterInfo.${responseField}`);
  }

```

I think it worth adding a `CfnOutput` for these attributes in the integ test, this would make sure the cr is invoked correctly.